### PR TITLE
8309067: gtest/AsyncLogGtest.java fails again in stderrOutput_vm

### PR DESCRIPTION
**Pull request by navyxliu**
This patch fixes the flaky test AsyncLogTest. The race condition only occurs in async 
logging mode in combination of BufferUpdater. BufferUpdater is introduced for testing 
only. It installs smaller buffers so test code can reliably observe messages dropped.
AsyncLog drops messages when the current buffer is overflown. 

In stdout/stderrOutput testcases, we used to mix test_asynclog_ls() and 
test_asynclog_drop_messages(). The later function uses BufferUpdater. As a result, log
messages distribute in both old buffer and new buffer. 

A race condition unfolds in the sequences. Thread names are in brackets.
1. [AsyncLogWriter] just calls write().
2. [Test] calls the constructor of BufferUpdater. It updates AsyncLogWriter->_buffer_staging.
3. [AsyncLogWriter] sees the latest _buffer_staging. It emits nothing. 
4. [Test] call the destructor of BufferUpdater. _buffer_staging resumes to 
the original buffer.
5. [AsyncLogWriter] is waken up from _lock.wait. It proceeds to swaps _buffer and _buffer_staging.
at this point, we lost the log messages in the original buffer. AsyncLogWriter was about to dump them
at 1.

Only stdout/stderrOutput testcases interleave the original buffer and small buffer. That's why
we only encounter flaky failures on them. It's up to the OS scheduling. 

To eliminate the race condition, we must ensure that all messages write after BufferUpdater.
test_asynclog_drop_messages() fuses test_asynclog_ls().

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8309067](https://bugs.openjdk.org/browse/JDK-8309067): gtest/AsyncLogGtest.java fails again in stderrOutput_vm (**Bug** - P3)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) ⚠️ Review applies to [1c1dca2e](https://git.openjdk.org/jdk/pull/16728/files/1c1dca2ea0d68a163651f723a55331456cdb993b)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16728/head:pull/16728` \
`$ git checkout pull/16728`

Update a local copy of the PR: \
`$ git checkout pull/16728` \
`$ git pull https://git.openjdk.org/jdk.git pull/16728/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16728`

View PR using the GUI difftool: \
`$ git pr show -t 16728`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16728.diff">https://git.openjdk.org/jdk/pull/16728.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16728#issuecomment-1818453680)

### CODE REVIEW

File: test/hotspot/gtest/logging/test_asynclog.cpp [hunk 1]
```diff
@@ -244,67 +244,68 @@ TEST_VM_F(AsyncLogTest, logBuffer) {
 }
 
 TEST_VM_F(AsyncLogTest, droppingMessage) {
+  if (AsyncLogWriter::instance() == nullptr)
+    return;
```
Comment (jdksjolen) [inline test/hotspot/gtest/logging/test_asynclog.cpp:248]: Please use braces, even for single statement if-bodies.

**Review by jdksjolen — APPROVED** [review]
Hi,

This looks good to me, happy to see that the issue was only in the testing code and not in production code.

File: test/hotspot/gtest/logging/test_asynclog.cpp (hunk 1 above)
Comment (navyxliu) [reply test/hotspot/gtest/logging/test_asynclog.cpp]: thanks. I resolve it.

**Review by dholmes-ora — APPROVED** [review]
Seems reasonable. Thanks for getting to the bottom of this. I too am glad to see it is a test issue.

**Pull request closed without merging**